modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdSelectivity.java

```diff
@@ -110,6 +122,654 @@ public Double getSelectivity(IgniteSortedIndexSpool rel, RelMetadataQuery mq, Re
+
+        SqlOperator op = ((RexCall)pred).op;
+
+        BigDecimal min = toComparableValue(colStat.min());
+        BigDecimal max = toComparableValue(colStat.max());
+        BigDecimal total = (min == null || max == null) ? null : max.subtract(min).abs();
+
+        if (total == null)
+            // No min/max mean that all values are null for coumn.
+            return 0.;
+
+        // All values the same so check condition and return all or nothing selectivity.
+        if (total.signum() == 0) {
+            BigDecimal diff = val.subtract(min);
+            int diffSign = diff.signum();
+
+            switch (op.getKind()) {
+                case GREATER_THAN:
+                    return (diffSign < 0) ? 1. : 0.;
+
+                case LESS_THAN:
+                    return (diffSign > 0) ? 1. : 0.;
+
+                case GREATER_THAN_OR_EQUAL:
+                    return (diffSign <= 0) ? 1. : 0.;
+
+                case LESS_THAN_OR_EQUAL:
+                    return (diffSign >= 0) ? 1. : 0.;
+
+                default:
+                    return guessSelectivity(pred);
+            }
+        }
+
+        // Estimate percent of selectivity by ranges.
+        BigDecimal actual = BigDecimal.ZERO;
+
+        switch (op.getKind()) {
+            case GREATER_THAN:
+            case GREATER_THAN_OR_EQUAL:
+                actual = max.subtract(val);
+
+                if (actual.signum() < 0)
+                    return 0.;
+
+                break;
+
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+                actual = val.subtract(min);
+
+                if (actual.signum() < 0)
+                    return 0.;
+
+                break;
+
+            default:
+                return guessSelectivity(pred);
+        }
+
+        return (actual.compareTo(total) > 0) ? 1 : actual.divide(total, MATH_CONTEXT).doubleValue();
+    }
+
+    /**
+     * Estimate "=" selectivity by column statistics.
+     *
+     * @param colStat Column statistics.
+     * @param comparableVal Comparable value to compare with.
+     * @return Selectivity.
+     */
+    private double estimateEqualsSelectivity(ColumnStatistics colStat, BigDecimal comparableVal) {
+        if (colStat.total() == 0)
+            return 1.;
+
+        if (colStat.total() - colStat.nulls() == 0)
+            return 0.;
+
+        if (colStat.min() != null) {
+            BigDecimal minComparable = toComparableValue(colStat.min());
+            if (minComparable != null && minComparable.compareTo(comparableVal) > 0)
+                return 0.;
+        }
+
+        if (colStat.max() != null) {
+            BigDecimal maxComparable = toComparableValue(colStat.max());
+            if (maxComparable != null && maxComparable.compareTo(comparableVal) < 0)
+                return 0.;
+        }
+
+        double expectedRows = (double)(colStat.distinct()) / (colStat.total() - colStat.nulls());
+
+        return expectedRows / colStat.total();
```

Comment:
let's assume there is a dataset of 1000 rows with only 2 distinct values and no nulls. So 
`expectedRows` = 2 / (1000 - 0) = 2 / 1000 = 0.002. According to the algorithm above the selectivity will be equals to `expectedRows` / `colStat.total()` = 0.002 / 1000 = 0.000002, whereas the real selectivity should be `0.5`